MM-60415 ignore deactivated bots

Willyfrog opened the pull request:
#### Summary

Ensure bots are not counted when looking at deactivated users.

To test this:
- create a bot
- deactivate the bot
- Either request a support packet and look at support_packet.yaml or use the [Get Analytics API](urhttps://api.mattermost.com/#tag/system/operation/GetAnalyticsOldl) **without** a team id

#### Ticket Link

[MM-60415](https://mattermost.atlassian.net/browse/MM-60415)

#### Screenshots

after | before looking at suport_packet.yaml

![CleanShot 2024-09-13 at 13 51 09@2x](https://github.com/user-attachments/assets/8a434750-0dc2-483c-ad33-5efa933f2401)

#### Release Note

```release-note
Ignore bots when counting deactivated accounts for statistics
```


[MM-60415]: https://mattermost.atlassian.net/browse/MM-60415?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

hanzei commented on server/channels/store/storetest/user_store.go:
```diff
@@ -4488,6 +4489,39 @@ func testUserStoreAnalyticsGetInactiveUsersCount(t *testing.T, rctx request.CTX,
+	u1.Email = MakeEmail()
+	_, err := ss.User().Save(rctx, u1)
+	require.NoError(t, err)
+	defer func() { require.NoError(t, ss.User().PermanentDelete(rctx, u1.Id)) }()
```
Nit: Use `t.Cleanup`

hanzei commented on server/channels/store/storetest/user_store.go:
```diff
@@ -4488,6 +4489,39 @@ func testUserStoreAnalyticsGetInactiveUsersCount(t *testing.T, rctx request.CTX,
 	require.Equal(t, count, newCount-1, "Expected 1 more inactive users but found otherwise.")
 }
 
+func testUserStoreAnalyticsGetInactiveUsersCountIgnoreBots(t *testing.T, rctx request.CTX, ss store.Store) {
```
0/5 This could also be a subtest of `testUserStoreAnalyticsGetInactiveUsersCount`.

Willyfrog replied:
I preferred to keep them separate, but I'm also not strongly oppossed

hanzei commented on server/channels/store/sqlstore/user_store.go:
```diff
@@ -1674,7 +1674,7 @@ func (us SqlUserStore) performSearch(query sq.SelectBuilder, term string, option
 func (us SqlUserStore) AnalyticsGetInactiveUsersCount() (int64, error) {
 	var count int64
-	err := us.GetReplicaX().Get(&count, "SELECT COUNT(Id) FROM Users WHERE DeleteAt > 0")
+	err := us.GetReplicaX().Get(&count, "SELECT COUNT(Id) FROM Users LEFT JOIN Bots ON Users.ID = Bots.UserId WHERE Users.DeleteAt > 0 AND Bots.UserId IS NULL")
```
This might be a good time to migrate the query to `squirrel` :smiley:

hanzei commented on server/channels/store/storetest/user_store.go:
```diff
@@ -4488,6 +4489,39 @@ func testUserStoreAnalyticsGetInactiveUsersCount(t *testing.T, rctx request.CTX,
+	require.NoError(t, err)
+	defer func() { require.NoError(t, ss.User().PermanentDelete(rctx, u1.Id)) }()
+
+	count, err := ss.User().AnalyticsGetInactiveUsersCount()
```
Should we check the value of `count`?

Willyfrog replied:
I would expect any problems with count to be detected by `AnalyticsGetInactiveUsersCount`

anything specific you wanted to test?

Willyfrog replied:
that's right, I've added it

hanzei reviewed (changes requested):
Nice work, just a few comments below

hanzei commented on server/channels/store/storetest/user_store.go:
```diff
@@ -4471,23 +4472,56 @@ func testUserStoreAnalyticsGetInactiveUsersCount(t *testing.T, rctx request.CTX,
 	count, err := ss.User().AnalyticsGetInactiveUsersCount()
 	require.NoError(t, err)
-
+	require.Equal(t, count, 0, "No users should have been inactive yet")
```
`require.Equal` is type sensitive :confused: 
```suggestion
	require.Equal(t, count, int64(0), "No users should have been inactive yet")

```

hanzei commented on server/channels/store/sqlstore/user_store.go:
```diff
@@ -1674,7 +1674,20 @@ func (us SqlUserStore) performSearch(query sq.SelectBuilder, term string, option
+		LeftJoin("Bots ON Users.ID = Bots.UserId").
+		Where(sq.And{
+			sq.Gt{"Users.DeleteAt": 0},
+			sq.NotEq{"Bots.UserId": nil},
```
The previous query was `SELECT COUNT(Id) FROM Users LEFT JOIN Bots ON Users.ID = Bots.UserId WHERE Users.DeleteAt > 0 AND Bots.UserId IS NULL`. Is it correct to check for `NotEq` here?

Willyfrog replied:
wow, that was a typo! thanks for catching it!

ifoukarakis commented on server/channels/store/sqlstore/user_store.go:
```diff
@@ -1674,7 +1674,20 @@ func (us SqlUserStore) performSearch(query sq.SelectBuilder, term string, option
+		LeftJoin("Bots ON Users.ID = Bots.UserId").
+		Where(sq.And{
+			sq.Gt{"Users.DeleteAt": 0},
+			sq.Eq{"Bots.UserId": nil},
```
Similar code in (https://github.com/mattermost/mattermost/blob/075681a412b56d177d54274f03852e4c93632f5a/server/channels/store/sqlstore/user_store.go#L1400) is using different query for postgres and mysql.  Perhaps the `Count` function can also be reused in order to keep logic in a single place?

Willyfrog replied:
I've changed the code to address the performance differences between Mysql and Postgresql

hanzei reviewed (approved):
LGTM :+1:

ifoukarakis reviewed (commented):
Left a minor comment. Not sure about the details as I'm not very familiar with the codebase.

hanzei reviewed (commented):
The code didn't change since my approval via https://github.com/mattermost/mattermost/pull/28184#pullrequestreview-2309975417.

Is there anything specific that I should check?

Willyfrog commented:
> The code didn't change since my approval via [#28184 (review)](https://github.com/mattermost/mattermost/pull/28184#pullrequestreview-2309975417).
> 
> Is there anything specific that I should check?

no, sorry! i removed an automatic review request that happened due to a rebase that went wrong and added a bunch of changes here. I fixed that and removed the automatic review request, and apparently that triggered a new request from you.

I'll add the suggestion above and then we can re-review it

Willyfrog commented:
@hanzei once the tests pass, you can re-review :)

hanzei commented on server/channels/store/sqlstore/user_store.go:
```diff
@@ -1674,7 +1674,27 @@ func (us SqlUserStore) performSearch(query sq.SelectBuilder, term string, option
 func (us SqlUserStore) AnalyticsGetInactiveUsersCount() (int64, error) {
 	var count int64
-	err := us.GetReplicaX().Get(&count, "SELECT COUNT(Id) FROM Users WHERE DeleteAt > 0")
+	query := us.getQueryBuilder().
+		Select("COUNT(Id)").
+		From("Users")
+	if us.DriverName() == model.DatabaseDriverPostgres {
+		query = query.LeftJoin("Bots ON Users.ID = Bots.UserId").
+			Where(sq.And{
+				sq.Gt{"Users.DeleteAt": 0},
+				sq.Eq{"Bots.UserId": nil},
+			})
+	} else {
+		query = query.Where(sq.And{
+			sq.Expr("Users.Id NOT IN (SELECT UserId FROM Bots)"),
+			sq.Gt{"Users.DeleteAt": 0},
+		})
+	}
+	queryStr, args, err := query.ToSql()
+	if err != nil {
+		return int64(0), errors.Wrap(err, "failed to create a SQL query to count inactive users")
+	}
+	err = us.GetReplicaX().Get(&count, queryStr, args...)
```
1/5 that the query separation is not worth it. MySQL is deprecated anyway, so the code won't be here for long. It's also a query that is not run very often, so I don't think the performance gains are worth it.

Willyfrog replied:
it is deprecated but we keep supporting it until v11, and on top of that there will be ESRs being supported for a while, so it is a longer period than a couple of months.

Now, for the performance gains, I'm not entirely sure how much gain do we get from this. The original effort was https://mattermost.atlassian.net/browse/MM-53438 , and it looks like a big gain, but again, this is an endpoint less used that the one on user's count

hanzei replied:
If we care about performance, we should  call `SqlUserStore.Count` as that method's performance has already been optimised.

hanzei replied:
Good point, I missed that `Count` doesn't support counting _only_ the inactive users.

Willyfrog replied:
I love to have this kind of conversations, forces me to explore further. Thanks!

Willyfrog commented:
/update-branch

hanzei reviewed (commented):
https://github.com/mattermost/mattermost/pull/28184#discussion_r1766383117 is still a concern for me

lindy65 reviewed (approved):
Thanks @Willyfrog - looks good to me 👍 Appreciate the help on how to get the support packet! I learnt something new today :)
